src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -364,152 +364,153 @@ void CoreWorkerDirectTaskSubmitter::RequestNewWorkerIfNeeded(
       /*grant_or_reject=*/is_spillback,
       [this, scheduling_key, task_id, is_spillback, raylet_address = *raylet_address](
           const Status &status, const rpc::RequestWorkerLeaseReply &reply) {
-        absl::MutexLock lock(&mu_);
+        std::deque<TaskSpecification> tasks_to_fail;
```

Comment:
is this file change purely adding error message?

Reply:
Ah no, there was actually a deadlock here before where the callback would call into the same class again and try to re-acquire the lock.

Reply:
Which class method will be called again inside the callback? Should we make the lock reentrant?

Reply:
SubmitTask, if the objects being failed here cause their dependent tasks to be resolved. IMO, I think this way of fixing the issue is more robust; if we allow the circular callback it can cause other internal data structures to get modified mid-execution.